File: frontend/api/tags.ts
import { apiClient } from "./client";
import type { Tag } from "./types";

export async function listTags(): Promise<Tag[]> {
  return (await apiClient.get<Tag[]>("/api/tags")).data;
}

export async function createTag(name: string, color: string | null = null): Promise<Tag> {
  return (await apiClient.post<Tag>("/api/tags", { name, color })).data;
}

export async function deleteTag(id: number): Promise<void> {
  await apiClient.delete(`/api/tags/${id}`);
}

File: backend/app/api/routes/tags.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.repositories import tag_repository
from app.schemas.tag import TagCreateRequest, TagResponse, tag_from_orm
from app.services.exceptions import NotFoundError

router = APIRouter(prefix="/api/tags", tags=["tags"], dependencies=[Depends(get_current_user)])


@router.get("", response_model=list[TagResponse])
def list_tags(db: Session = Depends(get_db)) -> list[TagResponse]:
    return [tag_from_orm(tag) for tag in tag_repository.list_tags(db)]


@router.post("", response_model=TagResponse, status_code=status.HTTP_201_CREATED)
def create_tag(body: TagCreateRequest, db: Session = Depends(get_db)) -> TagResponse:
    tag = tag_repository.get_or_create_by_name(db, body.name, body.color)
    db.commit()
    return tag_from_orm(tag)


@router.delete("/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_tag(tag_id: int, db: Session = Depends(get_db)) -> None:
    tag = tag_repository.get_tag(db, tag_id)
    if tag is None:
        raise NotFoundError("Tag not found")
    tag_repository.delete_tag(db, tag)
    db.commit()
